core-common/src/main/java/org/zalando/nakadi/util/MDCUtils.java

```diff
@@ -0,0 +1,114 @@
+package org.zalando.nakadi.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.Map;
+
+public final class MDCUtils {
```

Comment:
several concerns: 
- it is not clear why you need that code and can not use `MDC` class directly or via proxy
- the easy way to create just necessary resource like subscription id / stream / flow would be easier to read the code, instead you create the god method, which can do anything, is there a need in the future?
- the class creates hashmap per request plus MDC hashmap

Reply:
> it is not clear why that code is needed and can not use MDC class directly or via proxy

There is a lot of places that are using the same names and the same functionality, That is the reason why the code is concentrated in a single place

> the easy way to create just necessary resource like subscription id / stream / flow would be easier to read the code, instead it has a god method, which can do anything, is there a need in the future?

In order to push/pop several values and properly remove the values in case of unknown exceptions - there is a complex procedure to roll back.

> the class creates hashmap per request plus MDC hashmap

Thanks for spotting! I will do the optimization for the most obvious part - for 1 variable (flow id).

Reply:
After discussion - the complex procedure to roll back was removed and simplified. The second comment is obsolete now as well